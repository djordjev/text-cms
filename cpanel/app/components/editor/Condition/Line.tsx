import {
  IconSquareChevronRight,
  IconTrash,
  IconVariable
} from '@tabler/icons-react';
import classnames from 'classnames';
import { ChangeEvent, FC } from 'react';

import { ConditionDescriptor } from '~/types/condition';

export interface LineProps {
  className?: string;
  descriptor: ConditionDescriptor;
  index: number;
  onChange: (descriptor: ConditionDescriptor, index: number) => void;
  onDelete: (index: number) => void;
}

const Line: FC<LineProps> = (props) => {
  const { className, descriptor, index, onChange, onDelete } = props;

  // Setup
  const [variable, operator, value] = descriptor;

  // Styles
  const classes = classnames('u-flex u-gap-4x', className);

  // Handlers
  const onDeleteHandler = () => {
    onDelete(index);
  };

  const onVariableChange =
    (descriptorIndex: number) =>
    (e: ChangeEvent<HTMLSelectElement | HTMLInputElement>) => {
      const value = e.target.value;

      const newDescriptor: ConditionDescriptor = [...descriptor];
      newDescriptor[descriptorIndex] = value;

      onChange(newDescriptor, index);
    };

  return (
    <div className={classes}>
      <label className="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1">
        <IconVariable aria-label="Variable" className="u-mr-2x" />
        <input
          className="u-grow"
          onChange={onVariableChange(0)}
          placeholder="Enter variable name from payload"
          type="text"
          value={variable}
        />
      </label>

      <select
        aria-label="operator"
        className="u-select u-select-bordered"
        onChange={onVariableChange(1)}
        value={operator}
      >
        <option>=</option>
        <option>{'>'}</option>
        <option>{'>='}</option>
        <option>{'<'}</option>
        <option>{'<='}</option>
      </select>

      <label className="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1">
        <IconSquareChevronRight aria-label="Value" className="u-mr-2x" />
        <input
          className="u-grow"
          onChange={onVariableChange(2)}
          placeholder="Enter required value"
          type="text"
          value={value}
        />
      </label>

      <button aria-label="delete-line" onClick={onDeleteHandler} type="button">
        <IconTrash />
      </button>
    </div>
  );
};

export { Line };
