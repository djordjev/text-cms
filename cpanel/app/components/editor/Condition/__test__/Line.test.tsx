import { createRemixStub } from '@remix-run/testing';
import { fireEvent, render, screen } from '@testing-library/react';

import { Line, LineProps } from '../Line';

test('Line', () => {
  const props: LineProps = {
    descriptor: ['value', '=', 'on'],
    index: 2,
    onChange: vi.fn(),
    onDelete: vi.fn()
  };

  const createComponent = () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: () => <Line {...props} />
      }
    ]);

    return <RemixStub />;
  };

  const { container } = render(createComponent());

  // Renders all inputs
  expect(screen.getByRole('textbox', { name: 'Variable' })).toBeInTheDocument();
  expect(screen.getByRole('textbox', { name: 'Value' })).toBeInTheDocument();

  expect(screen.getByRole('option', { name: '=' })).toBeInTheDocument();
  expect(screen.getByRole('option', { name: '<' })).toBeInTheDocument();
  expect(screen.getByRole('option', { name: '<=' })).toBeInTheDocument();
  expect(screen.getByRole('option', { name: '>' })).toBeInTheDocument();
  expect(screen.getByRole('option', { name: '>=' })).toBeInTheDocument();

  expect(container.childNodes).toMatchSnapshot();

  // Calls onDelete
  fireEvent.click(screen.getByRole('button', { name: 'delete-line' }));
  expect(props.onDelete).toHaveBeenCalledOnce();
  expect(props.onDelete).toHaveBeenCalledWith(props.index);

  // Calls onChange for variable input
  const variableInput = screen.getByRole('textbox', { name: 'Variable' });
  fireEvent.change(variableInput, { target: { value: 'new-var' } });

  expect(props.onChange).toHaveBeenCalledOnce();
  expect(props.onChange).toHaveBeenCalledWith(
    ['new-var', '=', 'on'],
    props.index
  );

  // Calls onChange for value input
  vi.resetAllMocks();
  const valueInput = screen.getByRole('textbox', { name: 'Value' });
  fireEvent.change(valueInput, { target: { value: 'new-value' } });

  expect(props.onChange).toHaveBeenCalledOnce();
  expect(props.onChange).toHaveBeenCalledWith(
    ['value', '=', 'new-value'],
    props.index
  );

  // Calls onChange for operator input
  vi.resetAllMocks();
  const operatorInput = screen.getByRole('combobox');
  fireEvent.change(operatorInput, { target: { value: '>' } });

  expect(props.onChange).toHaveBeenCalledOnce();
  expect(props.onChange).toHaveBeenCalledWith(
    ['value', '>', 'on'],
    props.index
  );
});
