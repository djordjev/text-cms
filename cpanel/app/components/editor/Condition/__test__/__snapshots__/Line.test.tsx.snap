// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Line 1`] = `
NodeList [
  <div
    class="u-flex u-gap-4x"
  >
    <label
      class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
    >
      <svg
        aria-label="Variable"
        class="u-mr-2x"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 4c-2.5 5 -2.5 10 0 16m14 -16c2.5 5 2.5 10 0 16m-10 -11h1c1 0 1 1 2.016 3.527c.984 2.473 .984 3.473 1.984 3.473h1"
        />
        <path
          d="M8 16c1.5 0 3 -2 4 -3.5s2.5 -3.5 4 -3.5"
        />
      </svg>
      <input
        class="u-grow"
        placeholder="Enter variable name from payload"
        type="text"
        value="value"
      />
    </label>
    <select
      aria-label="operator"
      class="u-select u-select-bordered"
    >
      <option>
        =
      </option>
      <option>
        &gt;
      </option>
      <option>
        &gt;=
      </option>
      <option>
        &lt;
      </option>
      <option>
        &lt;=
      </option>
    </select>
    <label
      class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
    >
      <svg
        aria-label="Value"
        class="u-mr-2x"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 9l3 3l-3 3"
        />
        <path
          d="M3 3m0 2a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2z"
        />
      </svg>
      <input
        class="u-grow"
        placeholder="Enter required value"
        type="text"
        value="on"
      />
    </label>
    <button
      aria-label="delete-line"
      type="button"
    >
      <svg
        class="tabler-icon tabler-icon-trash"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 7l16 0"
        />
        <path
          d="M10 11l0 6"
        />
        <path
          d="M14 11l0 6"
        />
        <path
          d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
        />
        <path
          d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
        />
      </svg>
    </button>
  </div>,
]
`;
